test/test_torch.py

```diff
@@ -2195,16 +2195,14 @@ def test_device(self):
         self.assertEqual('cuda', cuda1.type)
         self.assertEqual(1, cuda1.index)
 
-        self.assertRaises(RuntimeError, lambda: torch.device('cpu:-1'))
```

Comment:
Why did you delete these?

Reply:
So in current master, c10/Device explicitly allows cpu with -1 index https://github.com/pytorch/pytorch/blob/dbf6d12776f87c878b46736c20d699b6e0d892e9/c10/Device.h#L43, what's the reason of it?

Reply:
The line you quoted is for the constructor. It clearly needs to support the -1 case because otherwise Device(kCPU) wouldn't work (due to the selection of default parameter.) That's fine. But it seems really hokey for that implementation detail to leak into the string parsing algorithm.

Reply:
No by "arbitrary" I didn't mean "random" at runtime, just not specifically hardcoded offline.
What does "cpu:0" mean currently in pytorch? Pin specifically to cpu socket 0 or whatever the "current socket" at runtime?

Reply:
Well, `cpu:0` doesn't mean anything in PyTorch, because we don't support NUMA. But `cuda:0` means specifically, put the CUDA memory on device 0. (And `cuda:-1` means, put it on the current CUDA device.)